File: sdk/i18n.py
import re
from pathlib import Path

import yaml


def _plural_form(n: int, locale: str) -> str:
  if locale == "ru":
    if n % 10 == 1 and n % 100 != 11:
      return "one"
    if n % 10 in (2, 3, 4) and n % 100 not in (12, 13, 14):
      return "few"
    return "many"
  return "one" if n == 1 else "other"


def _interpolate(template: str, vars: dict) -> str:
  return re.sub(r"\{\{(\w+)\}\}", lambda m: str(vars.get(m.group(1), m.group(0))), template)


class I18n:
  """Loads YAML translation files and resolves keys with variable substitution and pluralization."""

  def __init__(self, locale: str = "ru"):
    self._locale = locale
    self._data: dict = {}
    self._fallback: dict = {}

  def load_global(self, sdk_path: Path):
    self._load_into(sdk_path / "translations" / f"{self._locale}.yaml", self._data)
    if self._locale != "en":
      self._load_into(sdk_path / "translations" / "en.yaml", self._fallback)

  def load_plugin(self, plugin_path: Path):
    self._load_into(plugin_path / "translations" / f"{self._locale}.yaml", self._data)
    if self._locale != "en":
      self._load_into(plugin_path / "translations" / "en.yaml", self._fallback)

  def _load_into(self, path: Path, target: dict):
    if path.exists():
      with open(path, encoding="utf-8") as f:
        data = yaml.safe_load(f)
      if data:
        target.update(data)

  def t(self, key: str, **kwargs) -> str:
    val = self._data.get(key) or self._fallback.get(key) or key
    if isinstance(val, dict):
      val = val.get("other") or key
    return _interpolate(str(val), kwargs)

  def p(self, key: str, n: int, **kwargs) -> str:
    val = self._data.get(key) or self._fallback.get(key)
    if not isinstance(val, dict):
      return _interpolate(str(val or key), {"n": n, **kwargs})
    form = _plural_form(n, self._locale)
    template = val.get(form) or val.get("other") or key
    return _interpolate(template, {"n": n, **kwargs})

  def all(self) -> dict:
    result = dict(self._fallback)
    result.update(self._data)
    return result

  @property
  def locale(self) -> str:
    return self._locale

File: plugins/org_vrg_http/static/js/i18n.js
// Global i18n engine for VideoReg.
// Loaded before app.js; populated via VrgI18n.init() before Vue mounts.
const VrgI18n = (() => {
  let _locale = 'ru';
  let _data = {};

  function _pluralForm(n) {
    if (_locale === 'ru') {
      if (n % 10 === 1 && n % 100 !== 11) return 'one';
      if ([2, 3, 4].includes(n % 10) && ![12, 13, 14].includes(n % 100)) return 'few';
      return 'many';
    }
    return n === 1 ? 'one' : 'other';
  }

  function _interpolate(template, vars) {
    return template.replace(/\{\{(\w+)\}\}/g, (_, k) => (k in vars ? vars[k] : `{{${k}}}`));
  }

  return {
    async init() {
      try {
        const res = await fetch('/api/i18n', { credentials: 'same-origin' });
        if (res.ok) {
          const json = await res.json();
          _locale = json.locale || 'ru';
          _data = json.translations || {};
        }
      } catch (_) {}
    },

    t(key, vars = {}) {
      const val = _data[key];
      if (val == null) return key;
      const str = typeof val === 'object' ? (val.other ?? key) : val;
      return _interpolate(String(str), vars);
    },

    p(key, n, vars = {}) {
      const val = _data[key];
      if (val == null) return key;
      if (typeof val !== 'object') return _interpolate(String(val), { n, ...vars });
      const form = _pluralForm(n);
      const template = val[form] ?? val.other ?? key;
      return _interpolate(template, { n, ...vars });
    },

    get locale() { return _locale; },
  };
})();
